### PR TITLE
[BugFix] require gather for right outer join

### LINKED ISSUE #9583: Intrdouce the `NestLoopJoinOperator` in executor, make it support all kinds of join

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #9583 #10302

## Problem Summary(Required) ：

`RIGHT OUTER JOIN` needs to be gathered instead of broadcast to guarantee correctness.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test 
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 10 / 10 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|2|2|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 10 / 10 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java [hunk 1]
```diff
@@ -145,9 +145,15 @@ public Void visitPhysicalMergeJoin(PhysicalMergeJoinOperator node, ExpressionCon
 
     @Override
     public Void visitPhysicalNestLoopJoin(PhysicalNestLoopJoinOperator node, ExpressionContext context) {
-        PhysicalPropertySet rightBroadcastProperty =
-                new PhysicalPropertySet(new DistributionProperty(DistributionSpec.createReplicatedDistributionSpec()));
-        requiredProperties.add(Lists.newArrayList(PhysicalPropertySet.EMPTY, rightBroadcastProperty));
+        if (node.getJoinType().isRightJoin() || node.getJoinType().isFullOuterJoin()) {
+            PhysicalPropertySet rightGather =
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java:149]: add comment to descibe why right join need gather not broadcast in nest-loop?

Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java:151]: done

**Comment by mofeiatwork** [comment]
run all

**Review by Youngwb — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java [hunk 2]
```diff
@@ -145,9 +145,17 @@ public Void visitPhysicalMergeJoin(PhysicalMergeJoinOperator node, ExpressionCon
 
     @Override
     public Void visitPhysicalNestLoopJoin(PhysicalNestLoopJoinOperator node, ExpressionContext context) {
-        PhysicalPropertySet rightBroadcastProperty =
-                new PhysicalPropertySet(new DistributionProperty(DistributionSpec.createReplicatedDistributionSpec()));
-        requiredProperties.add(Lists.newArrayList(PhysicalPropertySet.EMPTY, rightBroadcastProperty));
+        if (node.getJoinType().isRightJoin() || node.getJoinType().isFullOuterJoin()) {
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java:148]: `onlyGather` is only appropriate to `NestloopJoinOperator`, not to other `JoinOperators`. So it depends on the physical join type, but not the logical join type. So I don't think the check should be add to `JoinHelper`.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 10 / 10 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|2|2|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 10 / 10 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|2|2|100.00%|[]|

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**